File: src/Topology.java
import java.util.*;

/*
    Class that finds different topological indexes
    basic topology -> frequency of valency pair
    ABC index -> sum over all edges(sqrt((du + dv - 2) / (du * dv))) where du and dv represent the degree of the vertices u, v of edge
*/

public class Topology {

    public static void traverseAndUpdateBasicTopology(Element currentElement, Set<Element> visited, HashMap<List<Integer>, Integer> frequencies) {

        visited.add(currentElement); // add the current carbon to visited hash map

        // a queue to store the carbons that must be visited next
        Queue<Element> toVisitCarbons = new LinkedList<>();

        Element[] connections = currentElement.getConnections(); // get all connections of current carbon

        // iterate through connections
        for (Element ithConnection : connections) {
            if (ithConnection != null) {
                // if the connection is a carbon, we append it to the queue to visit it later
                if (Objects.equals(ithConnection.getSymbol(), "C")) {
                    toVisitCarbons.offer(ithConnection);
                }

                // create a list of max size 2
                List<Integer> elementValencies = new ArrayList<>(2);

                /*
                 * NOTE
                 * valency pairs go both way, for example let us consider a bond C-H
                 * valency of carbon is 4 and valency of hydrogen is 1
                 * so two pairs can be produced from this bond alone which are as follows:
                 * pair1 = (4, 1) [valency of carbon, valency of hydrogen]
                 * pair2 = (1, 4) [valency of hydrogen, valency of carbon]
                 * thus, order of the valency matters
                 * however a pair of (1, 4) made from connection between chlorine and carbon and
                 * a pair of (1, 4) made from connection between hydrogen and carbon are equivalent
                 *
                 * the above stated rules apply for all bonds
                 */

                // place the carbon's valency and connection's valency and increment its frequency in the hashmap
                elementValencies.add(currentElement.getValency());
                elementValencies.add(ithConnection.getValency());
                Helper.incrementFrequency(frequencies, elementValencies);

                // place the valencies in reverse and increment its frequency in the hashmap
                elementValencies.set(0, ithConnection.getValency());
                elementValencies.set(1, currentElement.getValency());
                Helper.incrementFrequency(frequencies, elementValencies);
            }
        }

        // now we visit all carbons that is connected to the current carbon
        // if and only if the carbon is not already visited
        while(!toVisitCarbons.isEmpty()) {
            Element nextCarbon = toVisitCarbons.poll();
            if(!visited.contains(nextCarbon)) {
                traverseAndUpdateBasicTopology(nextCarbon, visited, frequencies);
            }
        }

    }


    /*
     * function to calculate the ABC (Atom-Bond Connectivity) index for the given compound.
     */
    public static double calculateABCIndex(Element[] compound) {

        // since abc index is a sum, we start the sum value at 0 for the result to not be biased (0 + number = number)
        double abcIndex = 0.0;

        // we iterate through each carbon of the compound
        for (Element element : compound) {
            Element[] connections = element.getConnections(); // get all the connections of the ith carbon
            int d_v = element.getNumConnections(); // get the degree of the ith carbon (number of connections of the ith carbon)

            // iterate through all connections of the ith carbon
            for (Element connectedElement : connections) {
                if (connectedElement == null) continue; // some connections can be just null, so skip it if it is

                // since we don't want to consider a C-C edge twice, we skip an edge if connection's element is lower than the current carbon,
                // since it is given that the left carbon has a lower ID than the right carbon
                if (connectedElement.getId() < element.getId()) continue;

                // get number of connections (degree) of the connection itself
                int d_w = connectedElement.getNumConnections();

                // if anyone is 0, we don't want to calculate the ABC index for that edge since it would cause a division by 0 error
                if (d_v == 0 || d_w == 0) {
                    System.out.println("Skipping edge with zero valency: " + d_v + ", " + d_w);
                    continue;
                }

                // calculate numerator
                double numerator = d_v + d_w - 2.0;
                // ABC index can't be negative, so if it goes below 0 we skip
                if (numerator < 0) {
                    System.out.println("Skipping edge due to negative numerator: " + d_v + ", " + d_w);
                    continue;
                }

                // find the ABC index for the current edge (term)
                double term = Math.sqrt(numerator / (d_v * d_w));

                // add the term to final result
                abcIndex += term;

            }
        }

        return abcIndex;

    }

}

File: topology.py
from typing import Tuple, DefaultDict
from element import Element
from collections import defaultdict, deque
import math

def find_basic_topology(ids_and_elements: DefaultDict[int, Element]) -> DefaultDict[Tuple[int, int], int]:

    # a set that we will use to not find basic topology for an element more than once
    visited = set()

    # a hashmap that says which valency pairs occured how many times
    basic_topology = defaultdict(int)
    
    starting_element = ids_and_elements[1] # start from the first element
    visited.add(starting_element) # add this element as visited first

    bfs_queue = deque() # bfs_queue, basically a queue to say which elements we should visit next
    bfs_queue.append(starting_element) # append the starting element to the queue

    while bfs_queue: # while there are elements in the queue
        ith_element = bfs_queue.popleft() # pop out an element

        for connection in ith_element.connections: # go through all connections
            if connection not in visited: # if connection has not already been seen before
                '''
                 * NOTE
                 * valency pairs go both way, for example let us consider a bond C-H
                 * valency of carbon is 4 and valency of hydrogen is 1
                 * so two pairs can be produced from this bond alone which are as follows:
                 * pair1 = (4, 1) [valency of carbon, valency of hydrogen]
                 * pair2 = (1, 4) [valency of hydrogen, valency of carbon]
                 * thus, order of the valency matters
                 * however a pair of (1, 4) made from connection between chlorine and carbon and
                 * a pair of (1, 4) made from connection between hydrogen and carbon are equivalent
                 *
                 * the above stated rules apply for all bonds
                '''

                element_valencies = [] # declare a list that will hold two numbers (valency pair)
                element_valencies.append(ith_element.valency) # append the valenciess
                element_valencies.append(connection.valency)

                basic_topology[tuple(element_valencies)] += 1 # increment frequency of that valency pair in the hash map

                element_valencies = element_valencies[::-1] # reverse the list
                basic_topology[tuple(element_valencies)] += 1 # increment frequency of the reversed valency pair in the hash map

                bfs_queue.append(connection) # append this connection to visit next
                visited.add(connection) # since we know we will visit this next, we can just add this to the visited set prematurely

    return dict(basic_topology) # return the basic topology hashmap


def compute_topological_indices(ids_and_elements: dict[int, Element], rr_alpha: float = 1.0) -> None: 
    # rr_alpha is a parameter used in calculating the Reciprocal Randić Index.
    # It defaults to 1.0 if not provided
    # None. This function only prints the calculated indices; it does not return them.


    visited_edges = set() # a set to keep track of visited edges to avoid double counting 
    M1 = M2 = HM = H = mm2 = ReZG3 = F = 0.0 # this will set the initial values to 0
    IS = A = R = RR = ABC = SC = 0.0

    
    for u in ids_and_elements.values(): # Iterate through all elements in the graph
        for v in u.connections: # For each connection of the current element
            if (u.id, v.id) in visited_edges or (v.id, u.id) in visited_edges: # Skip if this edge has already been processed
                continue

            du = len(u.connections) # number of connections of node u
            dv = len(v.connections) # number of connections of node v

            M1 += (du + dv) # First Zagreb Index: M1 = Σ (du + dv)
            M2 += (du * dv) # Second Zagreb Index: M2 = Σ (du * dv)
            HM += (du + dv) ** 2 # Hyper Zagreb Index: HM = Σ (du + dv)²
            H += 2 / (du + dv) # Harmonic Index: H = Σ 2 / (du + dv)
            mm2 += 1 / (du * dv) # Second Modified Zagreb Index: mm2 = Σ 1 / (du * dv)
            ReZG3 += (du * dv) * (du + dv) # Redefined Third Zagreb Index: ReZG3 = Σ (du * dv) * (du + dv)
            F += du ** 2 + dv ** 2 # Forgotten Index: F = Σ (du² + dv²)
            IS += (du * dv) / (du + dv) # Inverse Sum Indegree Index: IS = Σ (du * dv) / (du + dv)
            
            if du + dv - 2 != 0: #to avoid denominator being zero
                A += ((du * dv) / (du + dv - 2)) ** 3 # Augmented Zagreb Index: A = Σ ((du * dv) / (du + dv - 2))³
            
            R += 1 / math.sqrt(du * dv) # Randić Index: R = Σ 1 / √(du * dv)
            RR += 1 / ((du * dv) ** rr_alpha) # Reciprocal Randić Index: RR = Σ 1 / (du * dv) ^ α

            if du * dv != 0: #to avoid denominator being zero
                ABC += math.sqrt((du + dv - 2) / (du * dv)) # Atomic Bond Connectivity Index: ABC = Σ √((du + dv - 2) / (du * dv))
 
            SC += 1 / math.sqrt(du + dv) # Sum-connectivity Index: SC = Σ 1 / √(du + dv)

            visited_edges.add((u.id, v.id)) # Mark this edge as visited to avoid double counting

    print("\nTopological Indices:")
    print(f"First Zagreb Index (M1): {M1}") # First Zagreb Index
    print(f"Second Zagreb Index (M2): {M2}") # Second Zagreb Index
    print(f"Hyper Zagreb Index (HM): {HM}") # Hyper Zagreb Index
    print(f"Harmonic Index (H): {H:.4f}") # Harmonic Index
    print(f"Second Modified Zagreb Index (mm2): {mm2:.4f}") # Second Modified Zagreb Index
    print(f"Redefined Third Zagreb Index (ReZG3): {ReZG3}") # Redefined Third Zagreb Index
    print(f"Forgotten Index (F): {F}") # Forgotten Index
    print(f"Inverse Sum Indegree Index (IS): {IS:.4f}") # Inverse Sum Indegree Index
    print(f"Augmented Zagreb Index (A): {A:.4f}") # Augmented Zagreb Index
    print(f"Randic Index (R): {R:.4f}") # Randic Index
    print(f"Reciprocal Randic Index (RRalpha, alpha={rr_alpha}): {RR:.4f}") # Reciprocal Randic Index
    print(f"Atomic Bond Connectivity Index (ABC): {ABC:.4f}") # Atomic Bond Connectivity Index
    print(f"Sum-connectivity Index (SC): {SC:.4f}") # Sum-connectivity Index
